misc/python/materialize/cli/run.py

```diff
@@ -271,6 +273,39 @@ def main() -> int:
         raise UIError(f"unknown program {args.program}")
 
     print(f"$ {' '.join(command)}")
+
+    if args.program == "environmentd":
+        # Automatically restart `environmentd` after it halts, but not more than
+        # once every 5s to prevent hot loops. This simulates what happens when
+        # running in Kubernetes, which restarts failed `environmentd` process
+        # automatically. (We don't restart after a panic, since panics are
+        # generally unexpected and we don't want to inadvertently hide them
+        # during local development.)
+        while True:
+            last_start_time = datetime.now()
+            child_pid, exit_code = _launch(command, env)
+            if exit_code == 2:
+                wait = max(
+                    timedelta(seconds=5) - (datetime.now() - last_start_time),
+                    timedelta(seconds=0),
+                )
+                print(f"environmentd halted; will restart in {wait.total_seconds()}s")
+                time.sleep(wait.total_seconds())
+            else:
+                break
+    else:
+        child_pid, exit_code = _launch(command, env)
+
+    # Kill the child PID, which is the group into which any descendent processes
+    # spawned by the child process will exist in.
+    try:
+        os.killpg(child_pid, signal.SIGTERM)
```

Comment:
This approach is what I had as well, but doesn't that make it so we leak child processes once we go through a restart. The restarted processes will have a different child_pid, and it's child processes will be in a different group. So we don't kill them here.

I didn't test your PR locally, but that's what I got with a very similar (if not the same) approach.

Reply:
True! But those child processes should be `clusterd` processes that are managed by the process orchestrator, and thus readopted by the next `environmentd` incarnation. Though of course they'd be abandoned and leaked when the last `environmentd` incarnation exited. I was thinking we could live with that glitch but ...

...after thinking about this for a while I felt like the process group and session management here was overly complex. I've pushed up a new revision that vastly simplifies the management. I think it is nearly 100% robust and avoids the glitch you called out here. The only ways it can fail are if 1) if run.py itself crashes (unavoidable) or 2) if `environmentd` intentionally moves its subprocesses out into their own process group, which would be actively adversarial and something we don't need to worry about.